### PR TITLE
JupyterLab 3 support v2

### PR DESCRIPTION
**Pull request by krassowski**
Probably should have rebased rather than merging, but let's roll with this. Opening as new PR for now to spin the CI, but we might sent a PR to #402 instead. Relates to #400.

```
python -m pip install -e . --ignore-installed --no-deps -vv
jupyter serverextension enable --py jupyter_lsp
jupyter labextension develop . --overwrite
```

Punchlist:
- [x] setup new branch for JupyterLab 2.x and make RTD build another version of docs for it so that we can point users who will stick with 2.x for the time being (https://github.com/krassowski/jupyterlab-lsp/tree/2.x)
- [x] fix the docs
- [x] decide whether to publish both npm and python package or only one of those
    - probably npm is needed too as otherwise confused users will see message that this version of lab is not supported and we have no way of telling them from the interface that they need to use pip only this time.
- [x] typescript 4.1.3
- [x] eslint:
```
    "@typescript-eslint/eslint-plugin": "^4.8.1",
    "@typescript-eslint/parser": "^4.8.1",
    "eslint": "^7.14.0",
    "eslint-config-prettier": "^6.15.0",
    "eslint-plugin-jest": "^24.1.3",
    "eslint-plugin-prettier": "^3.1.4",
    "eslint-plugin-react": "^7.21.5",
```
- [x] In the CONTRIBUTING.md:
    - [x] the proposed `pip install .` needs to be invoked after every change, after compilation. This is extremely tedious and needs automation. [`jupyter labextension build` and `jupyter labextension develop . --overwrite`
](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#developing-a-prebuilt-extension) to the rescue?
- [ ] releasing instructions update
- [x]  update issue templates
- [ ] verify on [Binder](https://mybinder.org/v2/gh/krassowski/jupyterlab-lsp/jupyterlab3?urlpath=lab%2Ftree%2Fexamples%2FPython.ipynb)

### CODE REVIEW

**Comment by krassowski** [comment]
A small conda dependency hell for Python 3.6 it looks like. Did JupyterLab 3.0 drop 3.6?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -29,11 +29,11 @@ You can contribute to the project through:
 
 Development requires, at a minimum:
 
-- `nodejs >=10.12,<15`
+- `nodejs >=12,<15`
 - `python >=3.6,<3.9.0a0`
   - Python 3.6 and 3.8 are also tested on CI
```
Comment (krassowski) [inline CONTRIBUTING.md:34]:
```suggestion
  - Python 3.7 and 3.8 are also tested on CI
```

**Comment by krassowski** [comment]
Locally getting warnings:

- [ ] ServerApp] 'LanguageServerManager' object has no attribute 'virtual_documents_dir'
- [ ] jupyter_lsp | extension was found and enabled by nbclassic. Consider moving the extension to Jupyter Server's extension paths.

**Comment by krassowski** [comment]
- [ ] In the CONTRIBUTING.md:
    - [ ] the proposed `pip install .` needs to be invoked after every change, after compilation. This is extremely tedious and needs automation. [`jupyter labextension build` and `jupyter labextension develop . --overwrite`
](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#developing-a-prebuilt-extension) to the rescue?
- [ ] releasing instructions update

**Comment by bollwyvl** [comment]
In the `doit`-based automation I've been doing (ha) around lab3 on [other extensions](
https://github.com/jupyrdf/ipyelk/pull/62/files), the approach I'm somewhat happy with is:
- do a "normal" install by default (just to ensure it works)
- encapsulate the whole `watch` process in a separate command, out at the python level, which does the appropriate setup, runs the watches, etc. then kills both of them when either dies

I've been finding this easier to reason about.

On CI: node on windows (and even OSX) is boring when it breaks, as now we don't really care if webpack works on those platforms. I've been building the wheel/sdist up-front in a `build` job, and installing them in the `test` job. This helps catch subtle file omissions that the symlink thing sweeps under the covers. We may want a separate `package.json` of _just_ our npm-distributed language servers, which would be the only thing that's installed on the various test boxen, but wouldn't be installed on the `build` box. On a local dev enviornment, you'd need both, of course.

Testing: the modal command palette is kinda flaky under test (10% of the time, it seems to just disappear, depending on what has focus when invoked). The original sidebar behavior that is easy to test (and works with our existing stuff) can be restored by settings:

```robotframework
    Create File
    ...    ${SETTINGS DIR}${/}@jupyterlab${/}apputils-extension${/}palette.jupyterlab-settings
    ...    {"modal": false}
```

**Comment by bollwyvl** [comment]
> A small conda dependency hell for Python 3.6 it looks like. Did JupyterLab 3.0 drop 3.6?

this weirdness:
https://github.com/conda-forge/jupyter_server-feedstock/blob/master/recipe/meta.yaml#L19

Will get on it...

**Comment by bollwyvl** [comment]
Have some PRs up, one of which I can't merge (`jupyter_server`) but hopefully can be resolved quickly...

**Comment by krassowski** [comment]
I could not fix an issue with karma-webpack (tried updating to 5.0-alpha and updating everything, and tried few other workarounds):

```js
yarn run v1.21.1
$ karma start
02 01 2021 15:08:10.767:ERROR [preprocess]: Can not load "webpack", it is not registered!
  Perhaps you are missing some plugin?

START:
02 01 2021 15:08:10.803:ERROR [karma-server]: Error during file loading or preprocessing
TypeError: process is not a function
    at executeProcessor (/~/jupyterlab-lsp/node_modules/karma/lib/preprocessor.js:47:11)
    at runProcessors (/~/jupyterlab-lsp/node_modules/karma/lib/preprocessor.js:60:23)
    at FileList.preprocess [as _preprocess] (/~/jupyterlab-lsp/node_modules/karma/lib/preprocessor.js:134:11)
    at async Promise.all (index 0)
    at async /~/jupyterlab-lsp/node_modules/karma/lib/file-list.js:90:11
    at async Promise.all (index 2)
02 01 2021 15:08:10.804:INFO [karma-server]: Karma v5.2.3 server started at http://localhost:9876/
02 01 2021 15:08:10.804:INFO [launcher]: Launching browsers ChromeHeadless with concurrency unlimited
02 01 2021 15:08:10.805:ERROR [karma-server]: Error: Found 1 load error
    at Server.<anonymous> (/~/jupyterlab-lsp/node_modules/karma/lib/server.js:213:27)
    at Object.onceWrapper (events.js:421:28)
    at Server.emit (events.js:327:22)
    at emitListeningNT (net.js:1341:10)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
error Command failed with exit code 1.
```

Feel free to enable it if you have ideas how to make this work again. It would be a shame to trash the tests inherited from the original author (and I do not feel like rewriting them to jest right now).

**Comment by bollwyvl** [comment]
move to jest?

**Comment by krassowski** [comment]
Could it be that it uses wrong karma (should go for `packages/lsp-ws-connection/node_modules/karma` but uses `node_modules/karma`)? How would I fix that?

**Comment by bollwyvl** [comment]
Yeah, i don't have strong feelings on nodejs-based test stuff... i trust you've rebuilt the whole yarn.lock... are we pulling multiple versions? does it need `resolutions` of (more) things?

**Comment by krassowski** [comment]
It seems that it's just one version. I think I was wrong but on a good track:
- `karma` gets installed in `node_modules/`, but
- `karma-webpack` gets installed in `packages/lsp-ws-connection/node_modules`

so karma does not see the webpack plugin. Now I am not sure how to reconcile them.

**Comment by bollwyvl** [comment]
hoist karma-webpack to root `package.json`? there are some fiddly bits that can control it in a more fine-grained fashion but :shrug: ?

**Comment by bollwyvl** [comment]
and, of course, #278 dispenses with that package altogether...

**Comment by krassowski** [comment]
That's it from me for today. The CI is still red, but I was able to get it running locally. I think that I needed to enable serverextension with sys prefix in order for it to work, and I am still getting `extension was found and enabled by nbclassic` (for some reason `data_files` from setup do not end up installed when using `pip .` - I don't know where is problem).

**Comment by bollwyvl** [comment]
Awesome work! I'll give it some eyes. Don't sweat nbclassic stuff.
Data_files is a damn mess, not your fault. Like I said, if we run atest
against the real packages, we'll be able to make sure we don't need any
assets.

File: README.md [hunk 2]
```diff
@@ -77,9 +77,9 @@ Use context menu on rows in the panel to filter out diagnostics or copy their me
 
 You will need to have all of the following installed:
 
-- JupyterLab >=2.2.0,<3.0.0a0
+- JupyterLab >=3.0.0,<4.0.0a0
 - Python 3.6+
-- nodejs 10.12+
+- nodejs 12+
```
Comment (bollwyvl) [inline README.md:82]: Someone _just_ running `pyls` or `r-languageserver` won't need this at all :tada:!

File: packages/jupyterlab-lsp/package.json [hunk 3]
```diff
@@ -28,48 +28,63 @@
...
+    "clean": "jlpm run clean:lib",
+    "clean:all": "jlpm run clean:lib && jlpm run clean:labextension",
+    "clean:labextension": "rimraf labextension",
+    "clean:lib": "rimraf lib tsconfig.tsbuildinfo",
+    "eslint": "eslint . --ext .ts,.tsx --fix",
+    "eslint:check": "eslint . --ext .ts,.tsx",
+    "install:extension": "jupyter labextension develop --overwrite .",
     "lab:link": "jupyter labextension link . --no-build",
-    "test": "jlpm jest --coverage --coverageReporters=cobertura --coverageReporters=html --coverageReporters=text-summary"
+    "prepare": "jlpm run clean && jlpm run build:prod",
```
Comment (bollwyvl) [inline packages/jupyterlab-lsp/package.json:53]: this is one of those "magic" scripts, and gets run at "interesting" times... i recommend dropping it.

File: packages/jupyterlab-lsp/package.json [hunk 4]
```diff
@@ -28,48 +28,63 @@
...
-    "@jupyterlab/docmanager": "~2.2.0",
-    "@jupyterlab/docregistry": "~2.2.0",
-    "@jupyterlab/fileeditor": "~2.2.0",
-    "@jupyterlab/notebook": "~2.2.0",
-    "@jupyterlab/rendermime": "~2.2.0",
-    "@jupyterlab/services": "~5.2.0",
-    "@jupyterlab/statusbar": "~2.2.0",
-    "@jupyterlab/testutils": "~2.2.0",
-    "@jupyterlab/tooltip": "~2.2.0",
+    "@jupyterlab/application": "^3.0.0",
```
Comment (bollwyvl) [inline packages/jupyterlab-lsp/package.json:69]: we should probably hoist (at least) this to `dependencies`...

File: setup.py [hunk 5]
```diff
@@ -4,17 +4,26 @@
 
 import setuptools
 
+
+LABEXTENSIONS_DIR = Path('py_src') / 'jupyter_lsp' / 'labextensions'
+LABEXTENSIONS_INSTALL_DIR = Path('share') / 'jupyter' / 'labextensions'
+
+
+def get_data_files():
+    extension_files = [
```
Comment (bollwyvl) [inline setup.py:13]: `data_files` is an awful API. You have to give it each _directory_, so if there's any nesting, it doesn't do what one would expect. Here's [one approach](https://github.com/jupyrdf/ipyelk/pull/62/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7).

**Comment by bollwyvl** [comment]
Also, the upstream conda-forge PRs are coming along... should have something where we can drop the jedi pin, soon, and hopefully the `jupyter_server` one gets reviewed soon...

**Comment by bollwyvl** [comment]
At least part of the CI fail:

```
Run jupyter labextension install --debug --no-build packages/jupyterlab-lsp/krassowski-jupyterlab-lsp-2.1.2.tgz
```

Don't have to do this anymore :tada:, and even better, don't have to do `jupyter lab build` :partying_face: .

**Comment by krassowski** [comment]
So the key to enabling the extension with jupyter_server and not notebook is to add a space: `jupyter server extension enable` not `jupyter serverextension enable`. Subtle.

Also it turns out that installation from a wheel works ok with data, but `jupyter serverextension list` returns:

```
config dir: /~/.pyenv/versions/3.7.9/envs/jupyterlab-lsp-lab3/etc/jupyter
    jupyterlab  enabled 
    - Validating...
      jupyterlab 3.0.0 OK
```

whereas: `jupyter server extension list`:

```
Config dir: /~/.jupyter

Config dir: /~/.pyenv/versions/3.7.9/envs/jupyterlab-lsp-lab3/etc/jupyter
    jupyter_lsp enabled
    - Validating jupyter_lsp...
      jupyter_lsp 0.9.3 OK
    jupyterlab enabled
    - Validating jupyterlab...
      jupyterlab 3.0.0 OK
    nbclassic enabled
    - Validating nbclassic...
      nbclassic  OK

Config dir: /usr/local/etc/jupyter
```

So jupyterlab install itself to both notebook and server. The `jupyter_server` migration guide also [shows such an example](https://jupyter-server.readthedocs.io/en/latest/developers/extensions.html#migrating-an-extension-to-use-jupyter-server). However, lab extension cookicutter only enables [itself for server, not notebook](https://github.com/jupyterlab/extension-cookiecutter-ts/blob/master/%7B%7Bcookiecutter.python_name%7D%7D/setup.py).

I think that it should be ok.

New punchlist items:
- [ ] update issue template to use `jupyter server extension` for 3.0
- [ ] update manual enable in docs and readme to use `jupyter server extension` for 3.0

**Comment by bollwyvl** [comment]
Yeah, that's among the reasons I don't use the cookiecutter.

**Comment by bollwyvl** [comment]
Also, let's go ahead and bump `jupyter_lsp` to `1.0.0`, and add a hard dep on `jupyterlab >=3.0,<4`, I kinda don't trust the pip solver much these days, especially for upgrades...

**Comment by bollwyvl** [comment]
The work's paying off... my homebrew automation scripts have gotten me up to robot... it cracked a browser instead of using the headless one...

**Comment by krassowski** [comment]
Opened https://github.com/jupyterlab/jupyterlab/issues/9539 to as lab extension (by similarity to server extension) kind of works and sooner or later someone is gonna use it.

**Comment by bollwyvl** [comment]
Yeah, the problem is these are detected via `entry_points`, and it's bad business if two projects try to provide the same one...

**Comment by krassowski** [comment]
> it cracked a browser instead of using the headless one...

Got it, forgot to rename `jupyter_notebook_config.json` to server in one place

**Comment by krassowski** [comment]
So there is this failure:
```
Hover works in notebooks                                              | FAIL |
Keyword 'Trigger Via Hover With Modifier' failed after retrying 4 times. The last error was: WebDriverException: Message: TypeError: rect is undefined
------------------------------------------------------------------------------
Hover can be triggered via modifier key once cursor stopped moving    | FAIL |
Keyword 'Trigger Via Modifier Key Press' failed after retrying 5 times. The last error was: WebDriverException: Message: TypeError: rect is undefined
```

It would seem familiar for `Click Element`, but here it happens on `Mouse Over` and the element is definitely visible (except that `Element Is Not Visible` indicates its not). It is visible on screenshots and when debugging in Chrome but when debugging in robot it has size 0 x 0 **but** after painting blue rectangle over it it magically recovers its dimensions and works properly.

```robot
    ${width}  ${height} =   Get Element Size   ${sel}
    Run Keyword If  ${width} == 0   Cover Element   ${sel}    # don't know why but otherwise it raises Message: TypeError: rect is undefined
    Mouse Over    ${sel}
```

this is probably one of the weirdest workarounds I did ni a while, but it works. Any ideas what might be the root issue?

**Comment by bollwyvl** [comment]
that is a doozy... i am familiar with/dread the `rect is undefined` thing, and usually it's just a timing thing that needs longer. Seems to sometimes be related to react virtual rendering, etc. However, CodeMirror usually seems to have stuff on the page... i guess you could force a `refresh` or `repaint` or whatever with the CM API? At any rate, good to know about `Cover`.... very sneaky!

**Comment by bollwyvl** [comment]
Quick thing, probably need to add `py_src/jupyter_lsp/labextensions/` to `.gitignore`.

Less quick: I don't know the magic trick to make `jupyter_server` be the thing launched by binder.... will do some looking...

**Comment by krassowski** [comment]
> Quick thing, probably need to add py_src/jupyter_lsp/labextensions/ to .gitignore.

Already there.

Tests are now passing :tada: (well, not 3.6 but others do).

**Comment by krassowski** [comment]
@bollwyvl do you think that using `jupyter-lsp` as an umbrella for lab extension and server extension is ok, or should we try to split it into `jupyter-lsp` (server) and `jupyterlab-lsp` (lab)? I guess better to make this decision now rather than later.

I see both pros (one less thing to install, one less thing users will be confused about) and potential cons (in a distant future one may wish to provide lsp for other jupyter frontents and have frontend-specific packages).

**Comment by krassowski** [comment]
I think that having a separate `jupyterlab-lsp` (which depends on `jupyter-lsp`) will also avoid some confusion during upgrade for existing users. I now think that's the way to go.

**Comment by bollwyvl** [comment]
Yeah, two packages sounds good for now.

So maybe something like
- jupyterlab-lsp 3.0, requires
  - jupyter-lsp >=1.0.0,<2.0.0, requires
    - jupyter_server >=1.1.3
  - jupyterlab >=3.0.0,<4.0.0

Basic install instructions wouldn't even mention jupyter_lsp, probably.

**Comment by krassowski** [comment]
What about organization in repo:

```
/
|-- docs/
|-- packages/   <- typescript stuff
|-- python_packages/
   |-- jupyter_lsp/
   |  |-- src/*
   |  |-- setup.py
   |  `-- setup.cfg
   |-- jupyterlab_lsp/
   |  |-- src/
   |    |-- __init__.py
   |    `-- install.json
   |  |-- setup.py
   |  `-- setup.cfg
```

?

**Comment by bollwyvl** [comment]
Looks about right! I'm happy to take a crack at the re-org, if you're getting a little tired of the exercise: just let me know. As we hit more and more packages, I'm pining for `doit`...

**Comment by bollwyvl** [comment]
Also just got the first full local build of this through :tada: !

File: python_packages/jupyter_lsp/setup.py [hunk 6]
```diff
@@ -13,7 +13,7 @@
     data_files=[
         (
             "etc/jupyter/jupyter_server_config.d",
-            ["py_src/jupyter_lsp/etc/jupyter-lsp-serverextension.json"],
+            ["/etc/jupyter-lsp-serverextension.json"],
```
Comment (bollwyvl) [inline python_packages/jupyter_lsp/setup.py:16]: pretty sure it can't start with a `/`...

Comment (krassowski) [reply python_packages/jupyter_lsp/setup.py:16]: :laughing:

**Comment by krassowski** [comment]
The new jupyterlab-lsp package means that we need some work on conda side too - I will appreciate your help over there, I would be terribly slow learning how to do stuff properly on conda.

**Comment by bollwyvl** [comment]
Don't worry, we got you. Also: the layout should (unfortunately) be `python_packages/jupyter{,lab}_lsp/src/jupyter{,lab}_lsp`...

**Comment by krassowski** [comment]
Flag remained in setup.cfg, not sure what is going on:

```
python_packages/jupyter_lsp/src/non_blocking.py:12: error: Module 'ctypes' has no attribute 'WinError'
python_packages/jupyter_lsp/src/non_blocking.py:12: error: Module 'ctypes' has no attribute 'windll'
```

Removed flag jupyterlab on purpuse, I guess needs fixing because jupyter_lsp should not depend on lab:

```
python_packages/jupyter_lsp/src/types.py:244: error: Skipping analyzing 'jupyterlab': found module but no type hints or library stubs
```

**Comment by bollwyvl** [comment]
Yeah, see earlier comment: you're now shadowing the built-in `types`, etc. if you installed those into your pyenv... you're probably going to have to rebuild it :crying_cat_face:

**Comment by bollwyvl** [comment]
> `python_packages/jupyter_lsp/src/types.py:244: error: Skipping analyzing 'jupyterlab': found module but no type hints or library stubs`

That's probably for `pageConfig`?

**Comment by krassowski** [comment]
Ok, I got what you mean with organization. But nah, the ctypes should not be shadowed in that case...

**Comment by krassowski** [comment]
for node roots:

```python
    @default("node_roots")
    def _default_node_roots(self):
        """get the "usual suspects" for where `node_modules` may be found

        - where this was launch (usually the same as NotebookApp.notebook_dir)
        - the JupyterLab staging folder (if available)
        - wherever conda puts it
        - wherever some other conventions put it
        """

        # check where the server was started first
        roots = [pathlib.Path.cwd()]

        # try jupyterlab staging next
        try:
            from jupyterlab import commands

            roots += [pathlib.Path(commands.get_app_dir()) / "staging"]
        except ImportError:  # pragma: no cover
            pass

        # conda puts stuff in $PREFIX/lib on POSIX systems
        roots += [pathlib.Path(sys.prefix) / "lib"]

        # ... but right in %PREFIX% on nt
        roots += [pathlib.Path(sys.prefix)]

        return roots

```

**Comment by krassowski** [comment]
I guess it is fine, because it is optional...

**Comment by bollwyvl** [comment]
> for node roots:

Yeah, that's trash, and won't even generally be relevant anymore. Kill it with :fire: !

File: python_packages/jupyter_lsp/src/non_blocking.py [hunk 7]
```diff
@@ -9,8 +9,8 @@
 
 if os.name == "nt":  # pragma: no cover
     import msvcrt
-    from ctypes import POINTER, WinError, byref, windll, wintypes
-    from ctypes.wintypes import BOOL, DWORD, HANDLE
+    from ctypes import POINTER, WinError, byref, windll, wintypes  # type: ignore
```
Comment (bollwyvl) [inline python_packages/jupyter_lsp/src/non_blocking.py:12]: :+1:

**Comment by krassowski** [comment]
It looks (mostly) green again. Did you mean to send a PR against this branch rather than master? I would say safe to rebase now.

**Comment by bollwyvl** [comment]
I wanted to kick CI a few times before re-targeting this branch...

**Comment by krassowski** [comment]
Conda is fussy again `CondaHTTPError: HTTP 504 GATEWAY TIME-OUT for url <https://conda.anaconda.org/nodefaults/linux-64/current_repodata.json>`.

That's it from me for today.

**Comment by bollwyvl** [comment]
Again, heroic work. I'll see about setting up some higher retries and timeouts and stuff with a `.github/.condarc` over on #454, hopefully you'll have nice stuff to look at tomorrow.

**Comment by bollwyvl** [comment]
Py36 should be unblocked within the hour... If not, at least I can do
something directly about it now!

**Comment by bollwyvl** [comment]
hooray green. checking out binder (added to description).

> decide whether to publish both npm and python package or only one of those 

I feel it's important to keep publishing to `npm` so people can extend things. Let's not make another (worse) walled garden on pip.

**Comment by bollwyvl** [comment]
ugh, i forgot how _ugly_ it is if binder detects that you have r stuff, and goes off and does `apt` r, rstudio, shiny... I mean, it's all great stuff, but is not really germane to _our_ demo. I guess we change the name of our `environment.yml`, and run it our own selves in `postBuild`...

**Comment by bollwyvl** [comment]
Binder looks good enough! Since we don't do a `lab build` :tada: it's pretty snappy (once it gets to our stuff), but does pop the dreaded `build recommended` because of offline-notebook. Yeah, well, that ain't part of our demo either, so i think we just uninstall it.

Because it _does_ run, I think we hold off any heavy work on it.

Similarly, since we've got working CI, we could start a `develop` branch or something, make that a CI-enabled target, merge this there, and start working in parellel on smaller tasks with more meaningful diffs, as this is getting to be a bit of a monster. I just don't think we want to be blowing up RTD and the README.md with Instructions From The Future...

**Comment by krassowski** [comment]
Great to hear that! 3.0 is out and if the announcement gets out tomorrow then it would probably be fine to merge to master. I added link to RTD for 2.x in README; but indeed there are two things to resolve:
- publish on PyPI (should we publish rc first? may need versioning adjustments as npm uses different suffixes than PyPI)
- publish on conda-forge (does it need a PyPI package first?)

**Comment by bollwyvl** [comment]
> publish on conda-forge 

Again, don't worry, let's just get something out we feel good about that the interested _can_ install, through some means other than standing up our dev env. Unless we think we're going to split the repo, we can keep running it out of the existing feedstock... it's already complicated as hell (with the language servers, etc.) best we don't make things that need multiple PRs.

> (does it need a PyPI package first?)

Yeah, we _do_ need a pypi upstream sdist. It's kind of a _thing_ to get a custom `rc` "label" (like a sub-channel). However, if an CF rc helps us be more confident, we can certainly do it. But it's really no big deal if conda folk have to wait a sec, as folk generally don't use the `rc`s much, anyway, if you _do_ set it up.

> link to RTD for 2.x in README

Perhaps we call that `stable`? There's a lotta stuff in the RTD admin UI, but I know it's possible. I think it can build individual tags as well (you have to "activate" them or something). We're just fighting confusion and support issues from folks that don't read Big Bold Notes.

Anyhow, I'm happy to help out in any way (in addition to condachores, of course).

**Comment by krassowski** [comment]
So here is the release candidate 0 on PyPI to unblock the conda work and reserve name: https://pypi.org/project/jupyterlab-lsp/

**Comment by krassowski** [comment]
And the announcment of 3.0 went live about an one hour ago. I would say no need to get rc on conda, it's just good to have rc on PyPI. Let me know if I can help with the conda work and lets merge this to master and iterate on smaller things then.

**Comment by bollwyvl** [comment]
Oh, neat. Here's the process, I'll get on it: https://github.com/conda-forge/cfep/blob/master/cfep-05.md

**Comment by bollwyvl** [comment]
Anyhow, started framing it up, but not going to push too much harder on it this morning. LMK when there are releases, and I'll get on it!

**Comment by bollwyvl** [comment]
Also we want _some_ kind of test for the new package... just checking whether it's installed/the files exist/etc...

**Comment by martinRenou** [comment]
I completely missed this PR in my notifications, thanks a lot for the work! Tell me if you need anything :)

**Comment by krassowski** [comment]
Thank **you** for starting the effort well in advance.

**Comment by krassowski** [comment]
RC is live: https://pypi.org/project/jupyter-lsp/1.0.0rc0/

**Comment by bollwyvl** [comment]
Some notes:
- we should probably update the tagline
```
Multi-Language Server WebSocket proxy for Jupyter Notebook/Lab server
                                                          ^
                                                    Maybe just "Server"?
```

**Comment by bollwyvl** [comment]
`jupyter-lsp >1.0.0` does not match `1.0.0rc0`, which fails `pip check` (and likely means that someone would _not_ get anything if they tried to `pip install`)

- we should add this to `integrity`
- we should add `pip check` to CI

**Comment by bollwyvl** [comment]
- the `rc`s went out without tags
  - we need to decide how we're going to do tags in the future, e.g. jupyterlab-like
    - `v3.0.0` is
      - `release/jupyterlab-lsp/1.0.0`
      - `release/jupyter-lsp/2.0.0`
    - or give up, and do CalVer...

**Comment by bollwyvl** [comment]
https://github.com/conda-forge/jupyter-lsp-feedstock/pull/8

**Comment by bollwyvl** [comment]
Theoretically should work (within the hour):

[Binder](https://mybinder.org/v2/gist/bollwyvl/76a720a27be20aec815d171227dc28cd/HEAD?urlpath=lab%2Ftree%2Fenvironment.yml)

**Comment by bollwyvl** [comment]
Well, it built, but the serverextension is not enabled, because binder is running under `notebook` + `nbclassic`, not straight-up `jupyter_server`. I don't think this is going to be a long-term problem, so I'm willing to roll with it.

**Comment by krassowski** [comment]
Added integrity checks and removed rc tags. If it passes I will upload new version and merge.

Where would the `pip check` go?

**Comment by krassowski** [comment]
Added `pip check` to `job.test.yml` though earlier I was thinking about keeping it in build (which would require to install packages there too I guess)

**Comment by bollwyvl** [comment]
> Added pip check to job.test.yml

Yeah, sorry, mean to actually PR these things, got caught up in local election coverage :blush:.

Also, as we ran into a similar packaging issue over on `jupyterlab-classic` once it got to conda (this happens _all the time_), I PRd some stuff to do a relatively lightweight check over there... I can adapt it here.

Basic idea: do _just_ a run of the packages, as built, with _just_ `setup-python` (no node, conda) and do the CLI smoke tests, as they seem. Will get a PR up, but might not be able to finish it.

**Comment by bollwyvl** [comment]
#456 smoke tests are green, so probably ready to merge in here. Slowest one was 6 minutes, so not going to bother with caching pip.

**Comment by lresende** [comment]
Any eta for a working rc?

**Comment by krassowski** [comment]
1-2 hours for PyPI release, some time more until conda packages catch up. I will do one final push and publish the new release once CI passes.

**Comment by lresende** [comment]
Thank you @krassowski

**Comment by krassowski** [comment]
@lresende the release is fully out now. Please see: https://github.com/krassowski/jupyterlab-lsp/releases/tag/v3.0.0 for update instructions

**Pull request merged**